include/aspect/particle/generator/interface.h

```diff
@@ -0,0 +1,177 @@
+#define __aspect__particle_generator_interface_h
+
+#include <aspect/particle/particle.h>
+#include <aspect/plugins.h>
+#include <aspect/simulator_access.h>
+
+#include <deal.II/base/parameter_handler.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    /**
+     * A namespace in which we define everything that has to do with defining
+     * the particle generation.
+     *
+     * @ingroup ParticleGenerators
+     */
+    namespace Generator
+    {
+      /**
+       * Abstract base class used for classes that generate particles.
+       *
+       * @ingroup ParticleGenerators
+       */
+      template <int dim>
+      class Interface : public SimulatorAccess<dim>
+      {
+        public:
+          /**
+           * Constructor.
+           */
+          Interface() {}
+
+          /**
+           * Destructor. Made virtual so that derived classes can be created
+           * and destroyed through pointers to the base class.
+           */
+          virtual ~Interface () {}
+
+          /**
+           * Generate particles. Every derived class
+           * has to decide on the method and number of particles to generate,
+           * for example using input parameters declared in their
+           * declare_parameters and parse_parameters functions. This function
+           * should generate the particles and associate them to their according
+           * cells by inserting them into a multimap between cell and particle.
+           *
+           * @return A multimap containing cells and their contained particles.
+           */
+          virtual
+          std::multimap<types::LevelInd, Particle<dim> >
+          generate_particles() = 0;
+
+
+          /**
+           * Declare the parameters this class takes through input files. The
+           * default implementation of this function does not describe any
+           * parameters. Consequently, derived classes do not have to overload
+           * this function if they do not take any runtime parameters.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           * The default implementation of this function does not read any
+           * parameters. Consequently, derived classes do not have to overload
+           * this function if they do not take any runtime parameters.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        protected:
+          /**
+           * Generate a particle at the specified position and with the
+           * specified id. Many derived classes use this functionality,
+           * therefore it is implemented here to avoid duplication.
+           */
+          std::pair<types::LevelInd,Particle<dim> >
+          generate_particle(const Point<dim> &position,
+                            const types::particle_index id) const;
```

Comment:
I changed the function to throw a particular Exception when a point is not in the local subdomain. This way plugins can decide what to do (currently, most simply ignore this exception). This function itself is not a "collective" operation, it depends on the plugins that call it. It can be used to just generate particles in the local domain, or to generate all particles, and simply ignore the cases outside of the local domain, therefore I would not make the documentation more complicated here, but instead mention in the individual generators whether they are suited for a large amount of global particles (currently probability density function and random uniform distribution) or just for smaller numbers of particles (the others). Ascii file could be made more scalable by allowing to provide n_processes files, and letting every process read their own specific file, but this is something for a future work. I documented the exception though.